security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationResources.scala

```diff
@@ -46,18 +59,81 @@ trait AuthenticationResources extends Directives with FailFastCirceSupport {
   protected lazy val additionalRoute: Route = Directives.reject
 }
 
+object AuthenticationResources {
+
+  def apply(config: Config, classLoader: ClassLoader, sttpBackend: SttpBackend[Future, Any])(
+      implicit ec: ExecutionContext
+  ): AuthenticationResources = {
+    implicit val sttpBackendImplicit: SttpBackend[Future, Any] = sttpBackend
+    AuthenticationProvider(config, classLoader)
+      .createAuthenticationResources(config, classLoader)
+  }
+
+}
+
 trait AnonymousAccess extends Directives {
-  val anonymousUserRole: Option[String]
+  this: AuthenticationResources =>
+
+  protected def anonymousUserRole: Option[String]
+
+  protected def authenticateReally(): AuthenticationDirective[AuthenticatedUser]
+
+  protected def authenticateReally(credentials: PassedAuthCredentials): Future[Option[AuthenticatedUser]]
+
+  override protected final val authCredentialsCrypter: Crypter = AesCrypter
+
+  override def authenticationMethod(): EndpointInput[AuthCredentials] = {
+    anonymousUserRole match {
+      case Some(_) =>
+        // In WithAnonymousAccessFallback the encrypted password of the anonymous user is added as an Authorization
```

Comment:
In order in making this more understable let's:
1. Highlight that is the hack
2. Start with explanation of limitation of tapir
3. Use "fake Authorization header" to highlight that it isn't some kind of normal situation :)
4. Mention that we run some part of logic is done twice - the best with notice like `see withAnonymousAccessFallbackOnMissingCredentials` - the best would be to keep this part of code closer

Reply:
not relevant anymore